A whole bunch of this:
```
[2020-02-26T11:13:11.774Z] In file included from /home/ec2-user/workspace/test_rosdistro_fuse_PR-151/package/fuse_core/src/graph.cpp:34:0:
[2020-02-26T11:13:11.774Z] /home/ec2-user/workspace/test_rosdistro_fuse_PR-151/package/fuse_core/include/fuse_core/graph.h:379:47: error: invalid use of incomplete type ‘class ceres::Problem’
[2020-02-26T11:13:11.774Z]                          const ceres::Problem::EvaluateOptions& options = ceres::Problem::EvaluateOptions()) const = 0;
[2020-02-26T11:13:11.774Z]                                                ^~~~~~~~~~~~~~~
[2020-02-26T11:13:11.774Z] In file included from /home/ec2-user/workspace/test_rosdistro_fuse_PR-151/package/fuse_core/include/fuse_core/graph.h:48:0,
[2020-02-26T11:13:11.774Z]                  from /home/ec2-user/workspace/test_rosdistro_fuse_PR-151/package/fuse_core/src/graph.cpp:34:
[2020-02-26T11:13:11.774Z] /usr/include/ceres/covariance.h:43:7: note: forward declaration of ‘class ceres::Problem’
[2020-02-26T11:13:11.774Z]  class Problem;
[2020-02-26T11:13:11.774Z]        ^~~~~~~
[2020-02-26T11:13:11.774Z] In file included from /home/ec2-user/workspace/test_rosdistro_fuse_PR-151/package/fuse_core/src/graph.cpp:34:0:
```

I can paste the whole error output if you want, but it is this same error triggered from several different places in the code.

--- a/fuse_core/include/fuse_core/graph.h
+++ b/fuse_core/include/fuse_core/graph.h
@@ -46,6 +46,7 @@
 #include <boost/serialization/access.hpp>
 #include <boost/type_index/stl_type_index.hpp>
 #include <ceres/covariance.h>
+#include <ceres/problem.h>
 #include <ceres/solver.h>
 
 #include <ostream>
@@ -354,6 +355,29 @@ public:
    * @return            A Ceres Solver Summary structure containing information about the optimization process
    */
   virtual ceres::Solver::Summary optimize(const ceres::Solver::Options& options = ceres::Solver::Options()) = 0;
+
+  /**
+   * @brief Evalute the values of the current set of variables, given the current set of constraints.
+   *
+   * The values in the graph do not change after the call.
+   *
+   * If any of the output arguments is nullptr, it will not be evaluated. This mimics the ceres::Problem::Evaluate
+   * method API. Here all output arguments default to nullptr except for the cost.
+   *
+   * TODO(efernandez) support jacobian output argument
+   * The jacobian output argument is not exposed at the moment because its type is a CRSMatrix, that probably needs to
+   * be converted to another type.
+   *
+   * @param[out] cost      The cost of the entire problem represented by the graph.
+   * @param[out] residuals The residuals of all constraints.
+   * @param[out] gradient  The gradient for all constraints evaluated at the values of the current set of variables.
+   * @param[in]  options   An optional Ceres Problem::EvaluateOptions object that controls various aspects of the
+   *                       problem evaluation.
+   *                       See https://ceres-solver.googlesource.com/ceres-solver/+/master/include/ceres/problem.h#401
+   * @return True if the problem evaluation was successful; False, otherwise.
+   */
+  virtual bool evaluate(double* cost, std::vector<double>* residuals = nullptr, std::vector<double>* gradient = nullptr,
+                        const ceres::Problem::EvaluateOptions& options = ceres::Problem::EvaluateOptions()) const = 0;
 
   /**
    * @brief Print a human-readable description of the graph to the provided stream.

--- a/fuse_graphs/include/fuse_graphs/hash_graph.h
+++ b/fuse_graphs/include/fuse_graphs/hash_graph.h
@@ -309,6 +309,29 @@ public:
   ceres::Solver::Summary optimize(const ceres::Solver::Options& options = ceres::Solver::Options()) override;
 
   /**
+   * @brief Evalute the values of the current set of variables, given the current set of constraints.
+   *
+   * The values in the graph do not change after the call.
+   *
+   * If any of the output arguments is nullptr, it will not be evaluated. This mimics the ceres::Problem::Evaluate
+   * method API. Here all output arguments default to nullptr except for the cost.
+   *
+   * TODO(efernandez) support jacobian output argument
+   * The jacobian output argument is not exposed at the moment because its type is a CRSMatrix, that probably needs to
+   * be converted to another type.
+   *
+   * @param[out] cost      The cost of the entire problem represented by the graph.
+   * @param[out] residuals The residuals of all constraints.
+   * @param[out] gradient  The gradient for all constraints evaluated at the values of the current set of variables.
+   * @param[in]  options   An optional Ceres Problem::EvaluateOptions object that controls various aspects of the
+   *                       problem evaluation.
+   *                       See https://ceres-solver.googlesource.com/ceres-solver/+/master/include/ceres/problem.h#401
+   * @return True if the problem evaluation was successful; False, otherwise.
+   */
+  bool evaluate(double* cost, std::vector<double>* residuals = nullptr, std::vector<double>* gradient = nullptr,
+                const ceres::Problem::EvaluateOptions& options = ceres::Problem::EvaluateOptions()) const override;
+
+  /**
    * @brief Print a human-readable description of the graph to the provided stream.
    *
    * @param[out] stream The stream to write to. Defaults to stdout.

--- a/fuse_graphs/src/hash_graph.cpp
+++ b/fuse_graphs/src/hash_graph.cpp
@@ -414,6 +414,15 @@ ceres::Solver::Summary HashGraph::optimize(const ceres::Solver::Options& options
   return summary;
 }
 
+bool HashGraph::evaluate(double* cost, std::vector<double>* residuals, std::vector<double>* gradient,
+                         const ceres::Problem::EvaluateOptions& options) const
+{
+  ceres::Problem problem(problem_options_);
+  createProblem(problem);
+
+  return problem.Evaluate(options, cost, residuals, gradient, nullptr);
+}
+
 void HashGraph::print(std::ostream& stream) const
 {
   stream << "HashGraph\n"

